python/pyspark/sql/connect/session.py

```diff
@@ -418,6 +426,28 @@ def createDataFrame(
             # If no schema supplied by user then get the names of columns only
             if schema is None:
                 _cols = [str(x) if not isinstance(x, str) else x for x in data.columns]
+                infer_pandas_dict_as_map = (
+                    str(self.conf.get("spark.sql.execution.pandas.inferPandasDictAsMap")).lower()
+                    == "true"
+                )
+                if infer_pandas_dict_as_map:
+                    fields = []
+                    pa_schema = pa.Schema.from_pandas(data)
+                    spark_type: Union[MapType, DataType]
+                    for field in pa_schema:
+                        field_type = field.type
+                        if isinstance(field_type, pa.StructType):
+                            if len(field_type) == 0:
+                                raise PySparkValueError(
+                                    error_class="CANNOT_INFER_EMPTY_SCHEMA",
+                                    message_parameters={},
+                                )
+                            arrow_type = field_type.field(0).type
```

Comment:
This gets a type from the first field. Is this the same when users explicitly specify the schema? e.g., what happen if the dictionary contains different types of values?

Reply:
Let's at least add a test so the result is the same for both when the type is specified, and when it's not specified.

Reply:
if the dictionary contains different types of values, Arrow automatically tries converting the values in the same type and raise exception if it is not possible:

```python
>>> pdf = pd.DataFrame({"str_col": ['second'], "dict_col": [{"1": 0.7, "2": "1"}]})
>>> sdf = spark.createDataFrame(pdf)
Traceback (most recent call last):
  File "/Users/haejoon.lee/Desktop/git_repos/spark/python/pyspark/sql/pandas/serializers.py", line 302, in _create_array
    return pa.Array.from_pandas(
  File "pyarrow/array.pxi", line 1116, in pyarrow.lib.Array.from_pandas
  File "pyarrow/array.pxi", line 340, in pyarrow.lib.array
  File "pyarrow/array.pxi", line 86, in pyarrow.lib._ndarray_to_array
  File "pyarrow/error.pxi", line 91, in pyarrow.lib.check_status
pyarrow.lib.ArrowInvalid: Could not convert '1' with type str: tried to convert to double
```

Reply:
Tests added